sql/analyze.go

```diff
@@ -0,0 +1,1036 @@
+				return parser.DBool(true), nil
+			}
+			// x != y
+			return left, right
+		case parser.LE:
+			// a > x OR a <= y
+			if cmp != 1 {
+				// x = y
+				return parser.DBool(true), nil
+			}
+			// x != y
+			return left, right
+		}
+
+	case parser.GE:
+		switch rcmp.Operator {
+		case parser.EQ:
+			// a >= x OR a = y
+			if cmp != 1 {
+				// x >. y
+				return left, nil
+			}
+			// x < y
+			return left, right
+		case parser.NE:
+			// a >= x OR a != y
+			if cmp != 1 {
+				// x <= y
+				return parser.DBool(true), nil
+			}
+			// x > y
+			return right, nil
+		case parser.GT:
+			// a >= x OR a > y
+			if cmp != 1 {
+				// x <= y
+				return left, nil
+			}
+			// x > y
+			return right, nil
+		case parser.GE:
+			// a >= x OR a >= y
+			if cmp == -1 {
+				// x < y
+				return left, nil
+			}
+			// x >= y
+			return right, nil
+		case parser.LT, parser.LE:
+			// a >= x OR a < y
+			if cmp != 1 {
+				// x <= y
+				return parser.DBool(true), nil
+			}
+			// x > y
+			return left, right
+		}
+
+	case parser.LT:
+		switch rcmp.Operator {
+		case parser.EQ:
+			// a < x OR a = y
+			if cmp == 0 {
+				// x = y
+				return &parser.ComparisonExpr{
+					Operator: parser.LE,
+					Left:     lcmp.Left,
+					Right:    lcmp.Right,
+				}, nil
+			} else if cmp == 1 {
+				// x > y
+				return left, nil
+			}
+			// x < y
+			return left, right
+		case parser.NE:
+			// a < x OR a != y
+			if cmp == 1 {
+				return parser.DBool(true), nil
+			}
+			return right, nil
+		case parser.GT:
+			// a < x OR a > y
+			if cmp == 0 {
+				// x = y
+				return &parser.ComparisonExpr{
+					Operator: parser.NE,
+					Left:     lcmp.Left,
+					Right:    lcmp.Right,
+				}, nil
+			} else if cmp == 1 {
+				// x > y
+				return parser.DBool(true), nil
+			}
+			return left, right
+		case parser.GE:
+			// a < x OR a >= y
+			if cmp == -1 {
+				// x < y
+				return left, right
+			}
+			// x >= y
+			return parser.DBool(true), nil
+		case parser.LT, parser.LE:
+			// a < x OR (a < y OR a <= y)
+			if cmp == 1 {
+				// x > y
+				return left, nil
+			}
+			// x < y
+			return right, nil
+		}
+
+	case parser.LE:
+		switch rcmp.Operator {
+		case parser.EQ:
+			// a <= x OR a = y
+			if cmp == -1 {
+				// x < y
+				return left, right
+			}
+			// x >= y
+			return left, nil
+		case parser.NE:
+			// a <= x OR a != y
+			if cmp != -1 {
+				// x >= y
+				return parser.DBool(true), nil
+			}
+			// x < y
+			return right, nil
+		case parser.GT, parser.GE:
+			// a <= x OR (a > y OR a >= y)
+			if cmp != -1 {
+				// x >= y
+				return parser.DBool(true), nil
+			}
+			// x < y
+			return left, right
+		case parser.LT, parser.LE:
+			// a <= x OR a < y
+			if cmp == -1 {
+				// x < y
+				return right, nil
+			}
+			// x >= y
+			return left, nil
+		}
+	}
+
+	return parser.DBool(true), nil
+}
+
+func simplifyOneOrInExpr(left, right *parser.ComparisonExpr) (parser.Expr, parser.Expr) {
+	switch left.Operator {
+	case parser.EQ:
+		switch right.Operator {
```

Comment:
Yeah, I could perhaps make this into a `maybeSimplifyOneOrInExpr` that returns a boolean whether the signatures of the comparison expressions matched. But it seemed simpler to just have the caller do the pre-condition check. I suppose I could double-check and panic here. I'll do that in a separate cleanup PR as I've already started building another branch on top of this one.